### ISSUE TITLE
Getting code actions from ''Vetur', 'Eslint'' takes too long! 

### ISSUE THREAD

**Issue by daolanfler** [vuejs/vetur#2150]
- [x] I have searched through existing issues
- [x] I have read through [docs](https://vuejs.github.io/vetur)
- [x] I have read [FAQ](https://github.com/vuejs/vetur/blob/master/docs/FAQ.md)

## Info

- Platform: macOS
- Vetur version:0.26.1
- VS Code version: 1.47.3

## Problem

Since I upgraded Vetur from 0.25.0 to 0.26.1,  after I  edit and save .vue files, vscode get stuck with this promote: 
![image](https://user-images.githubusercontent.com/29777079/89968633-dd045b00-dc86-11ea-84b5-f5fffe8c09c1.png)
even though the change is very mirror. I have to downgrade to 0.25.0 to avoid this ! 

local eslint version is eslint@6.8.0 
here is my vsocde setting: 
```json
{
    "vetur.format.defaultFormatterOptions": {
        "prettier": {
            "semi": false,
            "singleQuote": true,
            "tabWidth": 4
        }
    },
    "vetur.validation.template": true,
    "vetur.format.options.tabSize": 4,
    "vetur.format.defaultFormatter.js": "prettier-eslint",
    "[vue]": {
        "editor.defaultFormatter": "octref.vetur"
    },
    "vetur.format.defaultFormatter.html": "prettier",
    "eslint.alwaysShowStatus": true,
    "eslint.validate": [
        "javascript",
        "javascriptreact",
        "vue",
        "typescript",
        "typescriptreact"
    ],
    "eslint.options": {
        "extensions": [
            ".html",
            ".js",
            ".vue",
            ".jsx"
        ]
    }
}
```

**Comment by shiyu3169** [comment]
I meet the same issue on Win as well. Downgrading to 0.26.0 temprarily solve problem

**Comment by AceYangWQ** [comment]
I have the same problem, too

**Comment by 3nuc** [comment]
Not sure if this is not another extension I have but just throwing it out there - does anybody's code sometimes get mangled (eg. one line gets duplicated) after Vetur + Eslint code actions finish after a long time?

**Comment by rs3d** [comment]
@3nuc This happens for me sometimes, when I cancel the long running action. Maybe eslint/prettier related?

**Comment by francoism90** [comment]
I'm having the same issue, any solutions/workarounds? :)

**Comment by octref** [comment]
- Does this reproduce when you only have Vetur extension enabled? (Especailly with ESLint disabled)
- Does setting `"eslint.codeActionsOnSave.mode": "problems"` help?

**Comment by francoism90** [comment]
@octref I will test and report back.

**Comment by sqal** [comment]
@octref 

> Does this reproduce when you only have Vetur extension enabled? (Especailly with ESLint disabled)

For me, yes. When I have both vetur and eslint enabled I am getting the same message as @daolanfler 

`Saving 'Test.vue': Getting code actions from ''Vetur', 'ESLint'' (configure)` 

and it hangs for 5 to 7 seconds. Disabling eslint doesn't fix it, Also, one important thing I have to add here is that the mentioned message is shown only once when I try to save .vue file with lint problems. After the message disappears, saving the same file or any other .vue file with lint problems is immediate. 

> Does setting "eslint.codeActionsOnSave.mode": "problems" help?

Nope.

**Comment by borie88** [comment]
I've had this issue for quite a long time. I'm using TS in a Quasar project and after a fresh open of VS Code and starting the dev server it works fine for a while. After about a half hour of use, it starts taking a very long time to finish linting lines before save. When saving, if the before-save linting didn't quite finish, I get the code actions pop up for up to half a minute. On a pretty powerful computer, so that shouldn't be the issue

**Comment by bbugh** [comment]
I'm not sure, but it seems to be related to this issue https://github.com/microsoft/vscode/issues/101555. My hunch is that incomplete save participants are hanging around in a process which ends up slowing everything down. I don't have any way to prove or reproduce, but they both seem to happen around the same time.

**Comment by francoism90** [comment]
@octref Just to confirm here the setting doesn't fix the 'Getting code actions from..' hang.

**Comment by Doeke** [comment]
Temporary workaround I use: `> Developer: Restart extension host`. After that linting/saving is fast but it slows down again over time. Must be something is causing the work for the linting process to pile up every time it runs? Issue also seems way more pronounced on my linux laptop than on my powerful windows desktop. I hope this gets fixed soon, it's been bugging me for way too long.

**Comment by imtiyazs** [comment]
> Temporary fix I use: `> Developer: Restart extension host`. After that linting/saving is fast but it slows down again over time. Must be something is causing the work for the linting process to pile up every time it runs? Based on how the highlighted errors change throughout the process, it's as if it's trying to lint every previously linted version of the file. Issue also seems way more pronounced on my linux laptop than on my powerful windows desktop. I hope this gets fixed soon, it's been bugging me for too long.

This doesn't work for me on mac :/

**Comment by octref** [comment]
I'll make a setting that disables `codeActions` in Vetur. Try it and let me know if the situation improves.

**Comment by imtiyazs** [comment]
> I'll make a setting that disables `codeActions` in Vetur. Try it and let me know if the situation improves.

Waiting up..

**Comment by elevatebart** [comment]
How about disabling eslint code action from eslint in Vue files? Keeping only vetur actions.

```json
{
"editor.codeActionsOnSave": {
    "source.fixAll.eslint": true
  },
  "[vue]": {
    "editor.codeActionsOnSave": {
      "source.fixAll.eslint": false
    }
  },
}
```

Would that help? I tried on my local, will report soon.

**Comment by elevatebart** [comment]
It seems to work and keep both extensions at bay.

**Comment by elevatebart** [comment]
Caveat: template autofix will not be proceeded.

**Comment by octref** [comment]
@elevatebart So in your case it's ESLint hanging up, not Vetur, is that correct?

**Comment by elevatebart** [comment]
It's the two waiting for each other I believe.

**Comment by imtiyazs** [comment]
May be this isn't the expected solution but I managed to fix it by resetting all the stuff.

1. Take backup of your settings.json
2. Remove all extensions and restart vscode (In my case, I removed Vetur, EsLint, Go, Beautify, TODO)
3. Install all the extensions again
4. Open User Settings -> Search for Formatters
<img width="600" alt="Screenshot 2020-09-13 at 12 12 47 PM" src="https://user-images.githubusercontent.com/31210898/93012044-71314d00-f5ba-11ea-82ad-8da424237163.png">

5. Select Vetur -> Format -> Default Formatter (HTML)===> Change default to prettier
<img width="634" alt="Screenshot 2020-09-13 at 12 13 18 PM" src="https://user-images.githubusercontent.com/31210898/93012053-83ab8680-f5ba-11ea-88fc-d5bdee1f5dc6.png">

6. Go to Prettier config and Make sure no boxes are checked
<img width="631" alt="Screenshot 2020-09-13 at 12 14 25 PM" src="https://user-images.githubusercontent.com/31210898/93012072-ab9aea00-f5ba-11ea-838e-2d3c1e6ee725.png">

and done !



Above steps fixed Vue Formatting for me,. (It does not get stuck on loading now)..

**Comment by francoism90** [comment]
@imtiyazs This doesn't fix the issue, I have reinstalled and cleaned my settings a few times already. The issues can be triggered when dealing with multiple formatting issues. Downgrading does make it trigger less, but still it seems to happen because some extensions seem to bit each other.

**Comment by imtiyazs** [comment]
> @imtiyazs This doesn't fix the issue, I have reinstalled and cleaned my settings a few times already. The issues can be triggered when dealing with multiple formatting issues. Downgrading does make it trigger less, but still it seems to happen because some extensions seem to waiting for each other.

Um.. Pretty hard to find exact scenario. I faced this issue in one of the repo. After switching the repo, the extensions worked perfectly (Without any change in settings). Later, I follow above steps for the same problematic repo and extensions start working perfectly like before..  

One important thing, I changed default formatters of Vetur that made it work perfectly. You can check above images 4,5 and 6.

**Comment by zhuqingguang** [comment]
Is there any resolution ? It's seems that my vscode is awayls being blocked by this.

**Comment by imtiyazs** [comment]
> Is there any resolution ? It's seems that my vscode is awayls being blocked by this.

Can you try these settings?

`{
  "workbench.sideBar.location": "right",
  "eslint.alwaysShowStatus": true,
  "editor.defaultFormatter": "esbenp.prettier-vscode",
  "html.format.contentUnformatted": "",
  "javascript.format.enable": false,
  "eslint.options": {
    "extensions": [".html", ".js", ".vue", ".jsx"]
  },
  "editor.codeActionsOnSave": {
    "source.fixAll.eslint": true
  },
  "editor.fontFamily": "Fira Code, Menlo, Monaco, 'Courier New', monospace",
  "editor.fontLigatures": true,
  "editor.fontSize": 14,
  "window.zoomLevel": 0,
  "vetur.format.defaultFormatter.js": "vscode-typescript",
  "vetur.format.defaultFormatter.html": "prettier",
  "javascript.format.insertSpaceBeforeFunctionParenthesis": true,
  "explorer.confirmDelete": false,
  "git.autofetch": true,
  "editor.formatOnSave": true
}`

**Comment by elevatebart** [comment]
with formatting ;):

```json
{
  "eslint.alwaysShowStatus": true,
  "editor.defaultFormatter": "esbenp.prettier-vscode",
  "html.format.contentUnformatted": "",
  "javascript.format.enable": false,
  "eslint.options": { "extensions": [".html", ".js", ".vue", ".jsx"] },
  "editor.codeActionsOnSave": { "source.fixAll.eslint": true },
  "editor.fontFamily": "Fira Code, Menlo, Monaco, 'Courier New', monospace",
  "editor.fontLigatures": true,
  "editor.fontSize": 14,
  "window.zoomLevel": 0,
  "vetur.format.defaultFormatter.js": "vscode-typescript",
  "vetur.format.defaultFormatter.html": "prettier",
  "javascript.format.insertSpaceBeforeFunctionParenthesis": true,
  "explorer.confirmDelete": false,
  "git.autofetch": true,
  "editor.formatOnSave": true
}
```

**Comment by zhuqingguang** [comment]
@elevatebart  @imtiyazs   thanks! I restart my mac and it's ok. Next time i will try the setting. 🤝🤝
But the strange thing is that i have more than one vscode running, but only one has problem and they use the same setting 😂😂

**Comment by imtiyazs** [comment]
> @elevatebart @imtiyazs thanks! I restart my mac and it's ok. Next time i will try the setting. 🤝🤝
> But the strange thing is that i have more than one vscode running, but only one has problem and they use the same setting 😂😂

Yea, that's what I mentioned few comment above.. The behaviour depends on your workspace or project folder too..

Someone gave me a hint about auto closing tags. That few auto closing tags might create issue for linters. However, didn't get any chance to test that.

**Comment by octref** [comment]
I added `vetur.languageFeatures.codeActions`. Try turning it off in the new version to see the issue still happens. If this is the cause I'll investigate further.

**Comment by zhuqingguang** [comment]
> I added `vetur.languageFeatures.codeActions`. Try turning it off in the new version to see the issue still happens. If this is the cause I'll investigate further.

hi,octref, thank you very much for your work.  But  i can't find the version of `0.28.0` in my vscode, is it published?

**Comment by imtiyazs** [comment]
> > I added `vetur.languageFeatures.codeActions`. Try turning it off in the new version to see the issue still happens. If this is the cause I'll investigate further.
> 
> hi,octref, thank you very much for your work. But i can't find the version of `0.28.0` in my vscode, is it published?

Same

**Comment by francoism90** [comment]
@octref Thanks, how does this work? :)
Just `vetur.languageFeatures.codeActions: false` or some other value?

**Comment by sqal** [comment]
```vetur.languageFeatures.codeActions: false``` unfortunately doesn't work for me, nothing has changed

**Comment by yoyo930021** [comment]
> `vetur.languageFeatures.codeActions: false` unfortunately doesn't work for me, nothing has changed

If you upgrade to `v0.28.0` and restart vscode?

**Comment by sqal** [comment]
Yes, I am using 0.28.0 and I restarted vscode before checking this setting. Are there any other steps/ settings I can enable to debug this issue? I tried to reproduce this issue in a newly created project but failed so far. I will keep trying.

**Comment by imtiyazs** [comment]
> vetur.languageFeatures.codeActions

Thanks!
It working good for me now. 

1. I set `"vetur.languageFeatures.codeActions": false` 
2. Changed default formatter to prettier-eslint
3. IT DIDNT WORK. GOT THE SAME ERROR.
4. I set `"vetur.languageFeatures.codeActions": true`  and again to false
5. THEN IT STARTED WORKING PERFECT  and now I am unable to reproduce it !

Lol. What's happening?

**Comment by gregveres** [comment]
The new version and disabling codeActions worked great for me. I now get instant intellisence and instant saving. Fantatistic!

**Comment by hendrikbunnick** [comment]
I get this issue right when Vetur is initializing in a large Project (eg. opening a .vue file), independent of Lint/codeAction Settings.
VueServerMain gets stuck at 99% CPU usage and the Vue Language Server does not shut down, when VSCode is terminated.

This issue started for me when updating from 0.25.0 to 0.26.0

A Workaround for me is:
1. Start VSCode on small Project or Single Vue File and let Vetur Initialize.
2. Add "Folder to Workspace" containing the Large Project

I usually cd into project directory and boot up vscode via "code ." in the project folder,
somehow vetur/vueServerMain seems to get overloaded this way.

**Comment by daolanfler** [comment]
In my case,  the repo is also a pretty big mono repo, in small projects, it works fine.

**Comment by dieegov** [comment]
The new version and disabling codeActions worked great!

Vetur version 0.28.0

**Comment by octref** [comment]
#2374 might have fixed this. Please try the new version 0.29.0 when it's out. If it's not fixed, feel free to open a new issue.

**Comment by francoism90** [comment]
@octref I haven't got the Eslint loading dialog in a while, seems to be fixed. :)

**Comment by gregveres** [comment]
> The new version and disabling codeActions worked great for me. I now get instant intellisence and instant saving. Fantatistic!

I know that I said this worked for me, but I stopped doing front end work for a while and focused on backend work. When I came back to the front end stuff this week, all Vue file editing was horrendous again. I am seeing a "saving" dialog on every save and the file is getting mangled as I edit. I have given up on intellisense and just typed away. When the system finally catches up, it has gone in and deleted chunks of the file and put in a bunch of semi colons etc. 

I am on 0.28.0 and I am using vetur.languageFeatures.codeActions: false. 

I will see if 0.29.0 fixes it.

**Comment by yoyo930021** [comment]
The v0.29.0 is released.
You can try it.

**Comment by gregveres** [comment]
It seems to be saving without the warning coming up and it doesn't seem to be corrupting the file. So it looks like it is fixed. I will keep an eye out over the next few days. Thank you very much for your work on Vetur!

**Comment by sqal** [comment]
0.29.0 nothing has changed for me :(

**Comment by hendrikbunnick** [comment]
0.29.0 does not help with the VueServerMain getting stuck at 99% CPU, as soon as vetur activates.
Im Still Stuck using the older versions.
:/

**Comment by sqal** [comment]
@hendrikbunnick may I ask which version works best for you?

**Comment by hendrikbunnick** [comment]
@sqal im using 0.25.0.

**Comment by yoyo930021** [comment]
> 0.29.0 does not help with the VueServerMain getting stuck at 99% CPU, as soon as vetur activates.
> Im Still Stuck using the older versions.
> :/

If you can provide a repro case, we can debug it.
Please read https://github.com/vuejs/vetur/blob/master/.github/NO_REPRO_CASE.md and open a new issue.

**Comment by shiyu3169** [comment]
0.29.0 doesn't work for me as well

**Comment by hendrikbunnick** [comment]
0.30.0 seems to be working for me.

**Comment by stijnjanmaat** [comment]
0.30.2 is not working for me, still the same issues as after 0.25.0

**Comment by yoyo930021** [comment]
> 0.30.2 is not working for me, still the same issues as after 0.25.0

https://github.com/vuejs/vetur/issues/2150#issuecomment-721039861

**Comment by MarvinRudolph** [comment]
I don't think it's a problem with Vetur but with VS Code. Getting the same popup every time I save and it's just a simple TypeScript & ESLint project.

**Comment by Christopher-Hayes** [comment]
I don't get this error message, but I came across this experiencing linting slowdowns on TS files. Fairly consistently as VSCodeVim user I'll add a type to an object, hit escape to move my cursor somewhere else in the file, but it'll instead freeze up for a half a second, and then spill the vim key bindings as text on the screen (ie `data: anyjjj`, where `j` is me trying to jump down a few lines.

**Comment by reifnotreef** [comment]
@MarvinRudolph there's this issue which has been open for a while as well:  
https://github.com/microsoft/vscode-eslint/issues/1055

**Comment by nicooprat** [comment]
ESlint fixes were instantaneous on every files except `.vue` ones. Solved it with this setting: `"vetur.validation.script": false,`

**Comment by rnenjoy** [comment]
Been working to fix this for 5 days now. Saving.. Getting code actions from "Vetur", "Eslint" on my monrepo takes alot of time.

[Trace - 13:38:41] Received response 'textDocument/codeAction - (7)' in 276674ms.
[Trace - 13:38:42] Sending request 'textDocument/formatting - (16)'.

[Trace - 13:39:50] Received response 'textDocument/hover - (8)' in 340344ms.
[Trace - 13:39:50] Received response 'textDocument/hover - (9)' in 340145ms.
[Trace - 13:39:50] Received response 'textDocument/hover - (10)' in 337802ms.
[Trace - 13:39:50] Received response 'textDocument/hover - (11)' in 337293ms.
[Trace - 13:39:50] Received response 'textDocument/hover - (12)' in 335211ms.
[Trace - 13:39:50] Received response 'textDocument/hover - (13)' in 332737ms.
[Trace - 13:39:50] Received response 'textDocument/hover - (14)' in 328641ms.
[Trace - 13:39:50] Received response 'textDocument/documentSymbol - (15)' in 118920ms.
[Trace - 13:39:52] Received response 'textDocument/formatting - (16)' in 70220ms.
[Trace - 13:39:52] Sending notification 'textDocument/didSave'.

**Comment by yoyo930021** [comment]
> Been working to fix this for 5 days now. Saving.. Getting code actions from "Vetur", "Eslint" on my monrepo takes alot of time.
> 
> [Trace - 13:38:41] Received response 'textDocument/codeAction - (7)' in 276674ms.
> [Trace - 13:38:42] Sending request 'textDocument/formatting - (16)'.
> 
> [Trace - 13:39:50] Received response 'textDocument/hover - (8)' in 340344ms.
> [Trace - 13:39:50] Received response 'textDocument/hover - (9)' in 340145ms.
> [Trace - 13:39:50] Received response 'textDocument/hover - (10)' in 337802ms.
> [Trace - 13:39:50] Received response 'textDocument/hover - (11)' in 337293ms.
> [Trace - 13:39:50] Received response 'textDocument/hover - (12)' in 335211ms.
> [Trace - 13:39:50] Received response 'textDocument/hover - (13)' in 332737ms.
> [Trace - 13:39:50] Received response 'textDocument/hover - (14)' in 328641ms.
> [Trace - 13:39:50] Received response 'textDocument/documentSymbol - (15)' in 118920ms.
> [Trace - 13:39:52] Received response 'textDocument/formatting - (16)' in 70220ms.
> [Trace - 13:39:52] Sending notification 'textDocument/didSave'.

Wow, is it a big project?
What size is your project?

**Comment by ninofiliu** [comment]
[I wrote a guide about a simple workaround](https://dev.to/ninofiliu/how-to-fix-vetur-taking-forever-to-save-a-file-127)

**Comment by guobing1993** [comment]
https://github.com/microsoft/vscode-eslint/issues/1055#issuecomment-862230316

**Comment by webdevnerdstuff** [comment]
Just starting having this same exact problem as of Sunday. Last Friday I did not have the problem.

**Comment by webdevnerdstuff** [comment]
Still having this problem. I tried to downgrade both Vetur and ESLint and still had the problem, but a tiny bit less. Then I decided to completely disable ESLint and remove all of its relevant settings from my settings file. The problem was just as bad. So I'm thinking this is definitely more a problem with Vetur than ESLint, because otherwise why would this error happen when ESLint is completely disabled?

@octref @yoyo930021 Is anybody even looking at this problem anymore? If you want a repo where I have been having this problem, I can provide you access to it. I already have ESLint looking at the same repo to try to problem solve on their side.

**Comment by webdevnerdstuff** [comment]
Is this project dead? Because it seems like nobody is responding to any issues, and there are a lot of open pull requests. What is going on?

**Comment by webdevnerdstuff** [comment]
It doesn't seem like anybody with Vetur is responding to any issues, comments, or pull-requests for quite a long time now. I'm not sure if this project is dead or not.

For anybody else still having this problem I put this in the ESLint Issue as a work around:

So after getting extremely frustrated at this error coming up constantly today preventing me from getting any work done, I spent all morning try to figure a way around getting it to stop. The ONLY thing that worked, or at least so far was uninstalling Vetur, removing all it's settings, and updated the default formatter for vue files to `vscode.typescript-language-features`.

So I had syntax highlighting I added this extension:
https://marketplace.visualstudio.com/items?itemName=jcbuisson.vue

So far, haven't seen the error since making these changes.

**Comment by webdevnerdstuff** [comment]
So out of frustration between constant errors, and authors of both extensions not responding to anybody (or any issues for that matter) for many weeks, I completely disabled Vetur and all of its settings were removed. I then installed other extension to get syntax highlighting and formatting. I'm still having the same problem showing up. 

That leads me to believe it's ESLint that's the problem, not Vetur. Perhaps Vetur has it's own problem with this as well, but I know for sure that with it turned off I still have the problem.

**Comment by francoism90** [comment]
I don't know if it helps, but it's indeed eslint taking a long time to process and not Vetur. After switching to TS, Airbnb and the latest depends the problem disappeared. 

I'm using Quasar and it seems their default present offers a good config when using Vetur as well.

**Comment by yoyo930021** [comment]
> Still having this problem. I tried to downgrade both Vetur and ESLint and still had the problem, but a tiny bit less. Then I decided to completely disable ESLint and remove all of its relevant settings from my settings file. The problem was just as bad. So I'm thinking this is definitely more a problem with Vetur than ESLint, because otherwise why would this error happen when ESLint is completely disabled?
> 
> @octref @yoyo930021 Is anybody even looking at this problem anymore? If you want a repo where I have been having this problem, I can provide you access to it. I already have ESLint looking at the same repo to try to problem solve on their side.

I can take a look.
In fact, so far I have not found any possible reasons.

**Comment by ninofiliu** [comment]
Eslint is not the issue, Vetur is, because when running things from the command line instead of asking vetur to manage stuff, things go fast

Here's a fix https://github.com/vuejs/vetur/issues/2150#issuecomment-828782749

**Comment by lolik9l** [comment]
The solution that worked for me:
I saw that I had an extension: "Vue Language Features (Volar)" and it had conflicts with the Vetur extension, disabling the latter helped.
![image](https://user-images.githubusercontent.com/30825379/220313100-1381d7a3-da06-4a76-8ce3-e503ff3a02bc.png)

**Comment by hghristov** [comment]
After updating to 0.37.2 it started happening to me too. Downgrading to 0.36.1 fixed it.

**Comment by parakawa** [comment]
I had the same problem, just reload the Vetur package (in the vscode extensions list) and it worked

**Comment by kazudeath1** [comment]
> After updating to 0.37.2 it started happening to me too. Downgrading to 0.36.1 fixed it.

Among the four Nuxt v2 repositories, this problem occurred in only one.
It was solved using the method above.

**Comment by mr-devboy** [comment]
Stuck with the same issue. Volar v1.2.0
